overlord/snapstate/backend/copydata.go

```diff
@@ -333,3 +333,70 @@ func (b Backend) RemoveExposedSnapHome(snapName string) error {
 
 	return firstErr
 }
+
+// XDGDirsUndoInfo contains information about what XDG dirs were moved or created
+// in InitXDGDirs. This allows the caller to undo the operation.
+type XDGDirsUndoInfo struct {
+	Moved   []string
+	Created []string
+}
+
+// InitXDGDirs renames .local/share, .config and .cache directories to their
+// post core22 migration locations. Directories that don't exist are created.
+// The returned XDGDirsUndoInfo struct can be used to undo this operation.
+func (b Backend) InitXDGDirs(info *snap.Info) (undoInfo *XDGDirsUndoInfo, err error) {
+	opts := &dirs.SnapDirOptions{HiddenSnapDataDir: true, MigratedToExposedHome: true}
+	users, err := allUsers(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	undoInfo = &XDGDirsUndoInfo{}
+
+	srcXdgDirs := []string{".config", ".cache", ".local/share"}
+	dstXdgDirs := []string{"xdg-config", "xdg-cache", "xdg-data"}
+
+	for _, usr := range users {
+		uid, gid, err := osutil.UidGid(usr)
+		if err != nil {
+			return undoInfo, err
+		}
+
+		revDir := info.UserDataDir(usr.HomeDir, opts)
+		for i, srcDir := range srcXdgDirs {
+			src := filepath.Join(revDir, srcDir)
+			dst := filepath.Join(revDir, dstXdgDirs[i])
+
+			if exists, _, err := osutil.DirExists(dst); err != nil {
+				return undoInfo, err
+			} else if exists {
+				return undoInfo, fmt.Errorf("cannot migrate XDG dir %q to %q because destination already exists", src, dst)
+			}
+
+			if exists, isDir, err := osutil.DirExists(src); err != nil {
+				return undoInfo, err
+			} else if exists && isDir {
+				if err := os.Rename(src, dst); err != nil {
+					return undoInfo, err
+				}
+
+				undoInfo.Moved = append(undoInfo.Moved, dst)
+
+				// since we copied the contents of the revisioned dir to ~/Snap previously,
+				// remove the XDG directories in there
+				exposedXdgDir := filepath.Join(info.UserExposedHomeDir(usr.HomeDir), srcDir)
```

Comment:
it's a bit strange to do this here, wouldn't it make more sense to ignore or remove this there when the actually copy happens. Also because from our discussion we might have to run InitXDGDirs even when we are doing the copying into ~/Snap again.